imap/http_jwt.c

```diff
@@ -0,0 +1,338 @@
+ *
+ * 3. The name "Carnegie Mellon University" must not be used to
+ *    endorse or promote products derived from this software without
+ *    prior written permission. For permission or any legal
+ *    details, please contact
+ *      Carnegie Mellon University
+ *      Center for Technology Transfer and Enterprise Creation
+ *      4615 Forbes Avenue
+ *      Suite 302
+ *      Pittsburgh, PA  15213
+ *      (412) 268-7393, fax: (412) 268-7395
+ *      innovation@andrew.cmu.edu
+ *
+ * 4. Redistributions of any form whatsoever must retain the following
+ *    acknowledgment:
+ *    "This product includes software developed by Computing Services
+ *     at Carnegie Mellon University (http://www.cmu.edu/computing/)."
+ *
+ * CARNEGIE MELLON UNIVERSITY DISCLAIMS ALL WARRANTIES WITH REGARD TO
+ * THIS SOFTWARE, INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
+ * AND FITNESS, IN NO EVENT SHALL CARNEGIE MELLON UNIVERSITY BE LIABLE
+ * FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN
+ * AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
+ * OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *
+ */
+
+#include <config.h>
+
+#include <string.h>
+#include <syslog.h>
+
+#include <sasl/saslutil.h>
+
+#include "assert.h"
+#include "global.h"
+#include "http_err.h"
+#include "util.h"
+
+#include "http_jwt.h"
+
+static int is_enabled = 0;
+
+static struct buf key = BUF_INITIALIZER;
+
+static time_t max_age = 0;
+
+struct jwt {
+    // Base64 parts
+    const char *joh;
+    size_t johlen;
+    const char *jws;
+    size_t jwslen;
+    const char *sig;
+    size_t siglen;
+
+    struct buf buf;
+};
+
+static inline int is_base64url_char(char c)
+{
+    return ((c >= '0' && c <= '9') ||
+            (c >= 'a' && c <= 'z') ||
+            (c >= 'A' && c <= 'Z') ||
+            (c == '-' || c == '_'));
+}
+
+HIDDEN int http_jwt_init(const char *keystr, int age)
+{
+    is_enabled = 0;
+
+    // Reset previous state
+    if (buf_len(&key)) {
+        size_t len = buf_len(&key);
+        char *oldkey = buf_release(&key);
+        memset(oldkey, 0, len);
+        free(oldkey);
+    }
+    max_age = 0;
+
+    if (!keystr) {
+        xsyslog(LOG_ERR, "Unexpected null key", NULL);
+        return HTTP_SERVER_ERROR;
+    }
+
+    if (strncmp(keystr, "HS256:", 6)) {
+        xsyslog(LOG_ERR, "Unexpected key algo specifier", "key=<%s>", keystr);
+        return HTTP_SERVER_ERROR;
+    }
+
+    if (charset_decode(&key, keystr + 6, strlen(keystr) - 6, ENCODING_BASE64URL)) {
+        xsyslog(LOG_ERR, "Invalid base64url key", "key=<%s>", keystr);
+        return HTTP_SERVER_ERROR;
+    }
+
+    static size_t HS256_MINLEN = 32; // see RFC 2104, section 3
+    static size_t HS256_MAXLEN = 64;
+
+    if (buf_len(&key) < HS256_MINLEN || buf_len(&key) > HS256_MAXLEN) {
+        xsyslog(LOG_ERR, "Key length outside allowed range", "range=[%zu:%zu]",
```

Comment:
It'd be nice to log the len we got, but we can live without it since it's in the config